### PR TITLE
Fix bug: party paid too much fee

### PR DESCRIPTION
**Pull request by bonomat**
Fixes two bugs: 
- one party pays too much
- overflow bug if collab closing with amount smaller than dust

Fixes https://github.com/itchysats/itchysats/issues/1996

### CODE REVIEW

File: src/protocol.rs [hunk 1]
```diff
@@ -469,10 +469,10 @@ impl Payout {
             }
             (false, true) => {
                 updated.maker_amount = Amount::ZERO;
-                updated.taker_amount = taker_amount - (fee + maker_amount);
+                updated.taker_amount = taker_amount - fee + maker_amount;
```
Comment (bonomat) [inline src/protocol.rs]: @luckysori : 
Note: this is a breaking change IF either party would receive 0.
I'm not sure we need to cover for this. Any thoughts?

Comment (luckysori) [reply src/protocol.rs]: What I understand is:

- Before, we gave the party who deserved the full total amount much less than what they were supposed to. Basically all liquidation CETs are broken.
- With these changes we fix this critical (but highly unlikely to affect us) bug.
- If the production maker uses this code against and old taker, they will always fail during contract setup because they will disagree on the payouts for the one liquidation CET.

Basically, it is hard to make changes to the protocol without breaking old clients. This change is correct in `maia`, but blindly applying it to `itchysats` will make previous versions incompatible. We should really consider staying compatible with old versions i.e. only deprecating this behaviour for the next version of `taker`.

File: src/protocol/transactions.rs [hunk 2]
```diff
@@ -513,16 +528,48 @@ impl Fee {
...
+/// We need to consider a few cases:
+/// - If both amounts are >= DUST, they share the fee equally
+/// - If one amount is < DUST, it set to 0 and the other output needs to cover for the fee.
+pub fn subtract_fee(
+    fee: Amount,
+    dust_limit_maker: Amount,
+    dust_limit_taker: Amount,
+    amount_maker: Amount,
+    amount_taker: Amount,
+) -> Result<(Amount, Amount)> {
```
Comment (luckysori) [inline src/protocol/transactions.rs:549]: 🔧 It seems like this belongs with the `Fee` type.

**Review by luckysori — APPROVED** [review]
Thank you for the extensive tests!

I think these changes are good, the challenge will be to deploy them in ItchySats without breaking all old versions.

File: src/protocol.rs (hunk 1 above)
Comment (bonomat) [reply src/protocol.rs]: I agree, how would you design backwards compatibility in this case?

Comment (luckysori) [reply src/protocol.rs]: Depend on two versions on `maia` and differentiate between an old version of the collaborative settlement protocol and the  new one.

Apparently one can depend on multiple versions of the same crate using the [rename-dependency cargo feature](https://stackoverflow.com/questions/58739075/how-do-i-import-multiple-versions-of-the-same-crate).

File: src/protocol/transactions.rs [hunk 3]
```diff
@@ -513,16 +528,48 @@ impl Fee {
         self.fee.ceil() as u64
     }
 
+    #[cfg(test)]
```
Comment (luckysori) [inline src/protocol/transactions.rs:531]: 🔧 You can get rid of `split`, it's only used in tests to test itself.

File: src/protocol/transactions.rs (hunk 2 above)
Comment (bonomat) [reply src/protocol/transactions.rs]: done :)

File: src/protocol/transactions.rs (hunk 3 above)
Comment (bonomat) [reply src/protocol/transactions.rs]: done

**Comment by bonomat** [comment]
I tried to rebase and move the refactoring up front but failed due to merge conflcts :/

**Review by luckysori — APPROVED** [review]
No worries about the order of the commits from my end. LGTM!

**Comment by bonomat** [comment]
No bors in this repository 😅
